lib/data/external/utils.js

```diff
@@ -15,6 +20,67 @@ const utils = {
     removeQuotes(word) {
         return word.slice(1, -1);
     },
+    // Before putting or deleting object on Azure, check if MPU exists with
+    // same key name. If it does, do not allow put or delete because Azure
+    // will delete all blocks with same key name
+    protectAzureBlocks(bucketName, objectKey, dataStoreName, log, cb) {
+        const mpuBucketName = `${constants.mpuBucketPrefix}${bucketName}`;
+        const splitter = constants.splitter;
+        const listingParams = {
+            prefix: `overview${splitter}${objectKey}`,
+            listingType: 'MPU',
+            splitter,
+            maxKeys: 1,
+        };
+
+        return metadata.listMultipartUploads(mpuBucketName, listingParams,
+        log, (err, mpuList) => {
+            if (err && !err.NoSuchBucket) {
+                log.error('Error listing MPUs for Azure delete',
+                    { error: err, dataStoreName });
+                return cb(errors.InternalError
+                    .customizeDescription('Error deleting object from ' +
+                    `Azure: ${err.message}`));
+            }
+            if (mpuList && mpuList.Uploads && mpuList.Uploads.length > 0) {
+                const error = errors.MPUinProgress;
+                log.error('Error deleting obj from Azure datastore',
+                    { error, dataStoreName });
+                return cb(error.customizeDescription('Cannot ' +
+                    'delete object from Azure data backend ' +
+                    'because MPU with same key name is ongoing.'));
+            }
+            // If listMultipartUploads returns a NoSuchBucket error or the
+            // mpu list is empty, there are no conflicting MPUs, so continue
+            return cb();
+        });
+    },
+    // user passes list of parts to commit, which need to be compared to parts
+    // stored in metadata to get part info
+    filterStoredKeys(storedParts, userPartList) {
+        const partNumbersToCommit = [];
+        userPartList.forEach(part => {
+            partNumbersToCommit.push(parseInt(part.PartNumber[0], 10));
+        });
+
+        const partsToCommit = [];
+        storedParts.forEach(part => {
+            // partLocations will always be an array of length 1
+            const partInfo = part.value.partLocations[0];
+            if (partNumbersToCommit.indexOf(partInfo.key) > -1) {
```

Comment:
isn't the partNumbersToCommit array an array of just numbers?

Reply:
yes, and the data that's stored as 'key' in our metadata is actually also the partNumber

Reply:
Instead of doing a filter for just azure here, let's do a part validation function early on in waterfall. Can skip for aws based on not having metadata parts.